
- для измененного варианта ВставкиДК_Родитель1-5_коррекция со штрихкодом (все подтягивания данных в 23-10-24Для коррекции вставок.ipynb)
- подтягивания поля вставки из файла ВставкиДК_Родитель1-5_коррекция
см readme
C:\Users\специалист ао\mypandas\work\2023\23-09-ХХ Переоценка ДК магазины выст\ReadMe2023-10-18.docx


# imports

In [58]:
import numpy as np
import pandas as pd
import re
import os

In [59]:
from rusprice.dictdiamonds import dcol_dia, dcla_dia, dpo_dia, dogr_dia_discount
#import rusprice.func_rpdiamonds_xlsx_v2022_2_CARtoKVG_v2 as func_rpdiamonds_v2
#func_rpdiamonds_xlsx_v2023_1_CARtoKVG.py
import rusprice.func_rpdiamonds_xlsx_v2023_1_CARtoKVG as func_rpdiamonds_v2

In [60]:
folder = '../'
#cm readme

folder_res = folder+'for_script/'
folder_res

'../for_script/'

In [61]:
import datetime
datetime.datetime.now().strftime('%d-%m-%Y_%H-%M')

'08-11-2023_14-13'

In [62]:
pd.options.display.float_format = '{:,.2f}'.format

# Входные данные

## из 1С

In [63]:
#1. Вводные/

In [64]:
os.listdir(folder+'1 Вводные/')

['23-01-01 Сводка ЮИ.xlsx',
 'выгрузка УТ 2023-08-10',
 'Выгрузка УТ 2023-09-19',
 'выгрузка УТ 2023-10-12',
 'выгрузка УТ 2023-10-15',
 'выгрузки из гиис',
 'Выгрузки ЮТД 2023-10-13',
 'Номенклатура код ЦБ по УТ из регистра сведений.xlsx',
 'Расчет 10-086 ИП Алексеев (3).xlsx',
 'расчет стоимости ю и  ИП Алексеев(акт№9 от 22 12 2021).xlsx']

In [65]:
files = os.listdir(folder+'1 Вводные/выгрузка УТ 2023-10-15')
files

['Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx',
 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx']

In [66]:
list(enumerate(files))

[(0, 'Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx'),
 (1, 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx')]

In [67]:
filename_z = files[0]
filename_r= files[1]
filename_z, filename_r

('Остатки1СУТ__все_склады_ДК_2023-10-15_закупочные.xlsx',
 'Остатки1СУТ__все_склады_ДК_2023-10-15_розница.xlsx')

In [68]:
df_z = pd.read_excel(folder+'1 Вводные/выгрузка УТ 2023-10-15/'+filename_z,header=9)#, dtype={5: str}

In [69]:
#Вывести итог таблицы
df_z.iloc[len(df_z)-1][[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']]

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
Name: 34286, dtype: object

на 15-10-2023 ут

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
Name: 34286, dtype: object
на 09-10-2023 ут
Количество конечный остаток         34334.0
Конечный остаток вес              122885.73
Оценка (конечный остаток)      883196152.97
Name: 34350, dtype: object

на 19-09-2023 ут?
Количество конечный остаток         34528.0
Конечный остаток вес              123441.14
Оценка (конечный остаток)      888714246.96
Name: 34544, dtype: object

In [70]:
df_z.dropna(subset=['Код'], inplace=True)

In [71]:
#Вывести сумм таблицы
df_z[[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']].sum()

Количество конечный остаток        34,269.00
Конечный остаток вес              122,668.35
Оценка (конечный остаток)     880,921,254.30
dtype: float64

на 13-10-2023 ут?
Количество конечный остаток    3.433400e+04
Конечный остаток вес           1.228857e+05
Оценка (конечный остаток)      8.831962e+08
dtype: float64
на 19-09-2023 ут?
Количество конечный остаток    3.452800e+04
Конечный остаток вес           1.234411e+05
Оценка (конечный остаток)      8.887142e+08
dtype: float64


In [72]:
# Загрузка файла с розничными ценами

In [73]:
df_r = pd.read_excel(folder+'1 Вводные/выгрузка УТ 2023-10-15/'+filename_r,header=9)#, dtype={5: str}

In [74]:
#Вывести итог таблицы
df_r.iloc[len(df_r)-1][[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']]

Количество конечный остаток          34,269.00
Конечный остаток вес                122,668.35
Оценка (конечный остаток)     2,908,177,532.89
Name: 34286, dtype: object

на 13-10-2023
Количество конечный остаток          34334.0
Конечный остаток вес               122885.73
Оценка (конечный остаток)      2915863314.89
Name: 34350, dtype: object

In [75]:
df_r.dropna(subset=['Код'], inplace=True)

In [76]:
#Вывести сумм таблицы
df_r[[ 'Количество конечный остаток', 'Конечный остаток вес',
       'Оценка (конечный остаток)']].sum()

Количество конечный остаток          34,269.00
Конечный остаток вес                122,668.35
Оценка (конечный остаток)     2,908,177,532.89
dtype: float64

Количество конечный остаток    3.433400e+04
Конечный остаток вес           1.228857e+05
Оценка (конечный остаток)      2.915863e+09
dtype: float64

на 19-09-2023 ут?
Количество конечный остаток    3.452800e+04
Конечный остаток вес           1.234411e+05
Оценка (конечный остаток)      2.935242e+09
dtype: float64

In [77]:
df_r.rename(columns={'Оценка (конечный остаток)':'розничная стоимость'}, inplace=True)

In [78]:
df_z.rename(columns={'Оценка (конечный остаток)':'закупочная стоимость'}, inplace=True)

In [79]:
#проверка соответствия кодов
len(df_r['Код'].unique()) ,len(df_r)

(34269, 34269)

In [80]:
len(df_z['Код'].unique()) ,len(df_z)

(34269, 34269)

In [81]:
set(df_z['Код'].unique()) == set(df_r['Код'].unique())

True

In [82]:
#Присоединяем к стоимости df_1c
df_1c = pd.merge(df_r,df_z[['Код', 'закупочная стоимость']]
                           ,how='left',right_on='Код',left_on='Код')

In [83]:
df_1c[['закупочная стоимость','розничная стоимость']].sum()

закупочная стоимость     880,921,254.30
розничная стоимость    2,908,177,532.89
dtype: float64

In [84]:
df_1c.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34269 entries, 0 to 34268
Data columns (total 38 columns):
 #   Column                                                                   Non-Null Count  Dtype  
---  ------                                                                   --------------  -----  
 0   Склад                                                                    0 non-null      object 
 1   Unnamed: 1                                                               0 non-null      float64
 2   Unnamed: 2                                                               0 non-null      float64
 3   Код                                                                      34269 non-null  object 
 4   Unnamed: 4                                                               0 non-null      float64
 5   Старый код (Справочник "Номенклатура" (Общие))                           28239 non-null  object 
 6   Склад.1                                                               

In [85]:
for x in ['Склад', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 4','Unnamed: 8','Склад.1','Unnamed: 33',  'Unnamed: 29', 'Unnamed: 30', 'Unnamed: 31','Unnamed: 32']:
    print(x, df_1c[x].unique()[:13])

Склад [nan]
Unnamed: 1 [nan]
Unnamed: 2 [nan]
Unnamed: 4 [nan]
Unnamed: 8 [nan]
Склад.1 ['Благовещенск' 'Выставочный' 'Интернет-магазин' 'Иркутск' 'Кружало'
 'Ленина, 28' 'Ленина23' 'Ленск' 'Мирный' 'Московское представительство'
 'Новосибирск' 'Пассаж' 'Склад "В пути"']
Unnamed: 33 [nan 'Сахаювелир ООО']
Unnamed: 29 ['Ювелирные изделия' nan 'Драгоценности Якутии ООО' 'МИЗО'
 'Сахаювелир ООО' 'Кристалл 2001  ООО' 'СУВЕНИРЫ' 'Золото Якутии ОАО АПЮК'
 'Разные поставщики' '0']
Unnamed: 30 ['Саха Минералс Лимитед' 'Ювелирные изделия' 'ИП "Алексеев В.М."'
 '№ 06-329/2 от 12.12.16' 'Драгоценности Якутии ООО' 'Сахаювелир ООО'
 '№03/10-212 от 05.11.07' '№6/2-08-КП от 14.01.2008г.' 'Невский Бриллиант'
 'Золото Якутии ОАО АПЮК' 'Сапфир М ООО' '№11/08-КП от 21.01.2008г.'
 'Покровский гранильный завод ООО']
Unnamed: 31 ['Sakha Minerals Limited SM21-01 от 28.09.2021' 'Из Москвы'
 '10-271 от 21.12.2018' '06-329/2 от 12.12.2016 Задание №4'
 '06-341/4 от 26.12.2016' '03/10-95 от 15.05.08' '89/1/06 от 

In [86]:
df_1c.drop(['Склад', 'Unnamed: 1', 'Unnamed: 2','Unnamed: 4','Unnamed: 8'], axis=1, inplace=True)

In [87]:
df_1c.rename(columns={'Склад.1':'Склад','Unnamed: 29':'родитель1_1c', 'Unnamed: 30':'родитель2_1c',
                      'Unnamed: 31':'родитель3_1c', 'Unnamed: 32':'родитель4_1c', 'Unnamed: 33':'родитель5_1c'},
            inplace=True)

In [88]:
df_1c

Код Старый код (Справочник "Номенклатура" (Общие))  \
0      ЦБ-00200503                                            NaN   
1      ЦБ-00002854                                     МС00057368   
2      ЦБ-00019951                                     МС00152100   
3      ЦБ-00019953                                     МС00146311   
4      ЦБ-00111074                                     МС00111422   
...            ...                                            ...   
34264  ЦБ-00182016                                     0000032928   
34265  ЦБ-00182017                                     0000032929   
34266  ЦБ-00182019                                     0000032934   
34267  ЦБ-00182020                                     0000032935   
34268  ЦБ-00182021                                     0000032936   

                   Склад Склад.Склад / Магазин  \
0           Благовещенск     Розничный магазин   
1           Благовещенск     Розничный магазин   
2           Благовещенск     Розничный магазин   
3           Благовещенск     Розничный магазин   
4           Благовещенск     Розничный магазин   
...                  ...                   ...   
34264  Ювелирные изделия         Оптовый склад   
34265  Ювелирные изделия         Оптовый склад   
34266  Ювелирные изделия         Оптовый склад   
34267  Ювелирные изделия         Оптовый склад   
34268  Ювелирные изделия         Оптовый склад   

             Вставки (Справочник "Номенклатура" (Общие)) Вид номенклатуры  \
0                                      45КР57-0,520-4/5А     Т ДМ Юи с ДК   
1      2Кр57-0,083-2/2А 4Кр57-0,084-2/2А 6Кр57-0,280-...     Т ДМ Юи с ДК   
2                                       1Кр57-0,086-4/5А     П ДМ Юи с ДК   
3                                        3Кр57-0,06-4/5А     П ДМ Юи с ДК   
4      1Кр57-0,050-4/5А 1Кр57-0,031-6/6А 1Кр57-0,014-...     Т ДМ Юи с ДК   
...                                                  ...              ...   
34264                                                NaN     Т ДМ Юи с ДК   
34265                                                NaN     Т ДМ Юи с ДК   
34266                                                NaN     Т ДМ Юи с ДК   
34267                                                NaN     Т ДМ Юи с ДК   
34268                                                NaN     Т ДМ Юи с ДК   

      Тип номенклатуры Номенклатура                           Группа  \
0                Товар      Браслет  спецификация №1 от 27.01.2022г.   
1                Товар      Браслет                         Товар ДК   
2                Товар        Брошь                            Броши   
3                Товар        Брошь                          накл.№5   
4                Товар        Брошь                            51 ДК   
...                ...          ...                              ...   
34264            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34265            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34266            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34267            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   
34268            Товар  серьги с ДК        Акт №1 АК "Золото Якутии"   

      Заводской номер (Справочник "Номенклатура" (Общие))  ...  \
0                                                    512   ...   
1                                               30575,51   ...   
2                                                1191890   ...   
3                                                   1094   ...   
4                                                    422   ...   
...                                                  ...   ...   
34264                                           01-12308   ...   
34265                                           01-12309   ...   
34266                                           01-14045   ...   
34267                                           01-14046   ...   
34268                                           01-14047   ...   

  

In [89]:
# !!!
df_1c_res = df_1c

# Подтянуть вставки  ШК, родитель 1 -5 из файла ВставкиДК_Родитель1-5_коррекцияНазовем их Вставки_корр

Для корректировки вставок и р1-5 Берем из файла - C:\Users\специалист ао\mypandas\work\2023\23-09-ХХ Переоценка ДК магазины выст\Вставки Родитель1-5 коррекция\ВставкиДК_Родитель1-5_коррекция_18-10-2023_10-00.xlsx
Взамен этого файла Данные по вставкам надежнее чем с 1с (на 1 янв 2023г) \\UR-JVV\Users\DAS\Desktop\GM TMP\Analytics\Проекты в работе\23-09-XX Переоценка Самородки\1. Вводные\Данные по вставкам\23-01-01 Сводка ЮИ.xlsx

In [90]:
#'C:/Users/специалист ао/mypandas/work/2023/23-09-ХХ Переоценка ДК магазины выст/Вставки Родитель1-5 коррекция/ВставкиДК_Родитель1-5_коррекция_18-10-2023_10-00.xlsx'
files = os.listdir('C:/Users/специалист ао/mypandas/work/2023/23-09-ХХ Переоценка ДК магазины выст/Вставки Родитель1-5 коррекция/')
files

['~$ВставкиДК_Родитель1-5_коррекция_071120231743.xlsx',
 'архив',
 'ВставкиДК_Родитель1-5_коррекция_071120231743.xlsx',
 'Запасные копии - не изменять',
 'Корректировки от 23-24-окт 2023',
 'Материалы']

In [91]:
list(enumerate(files))

[(0, '~$ВставкиДК_Родитель1-5_коррекция_071120231743.xlsx'),
 (1, 'архив'),
 (2, 'ВставкиДК_Родитель1-5_коррекция_071120231743.xlsx'),
 (3, 'Запасные копии - не изменять'),
 (4, 'Корректировки от 23-24-окт 2023'),
 (5, 'Материалы')]

In [93]:
fn = files[2]
fn

'ВставкиДК_Родитель1-5_коррекция_071120231743.xlsx'

In [94]:
df_vst = pd.read_excel('C:/Users/специалист ао/mypandas/work/2023/23-09-ХХ Переоценка ДК магазины выст/Вставки Родитель1-5 коррекция/'\
                       +fn)#,header=9)#, dtype={5: str}

In [95]:
df_vst.columns

Index(['копия_', 'к_Склад', 'копия_.1', 'к_Штрихкод', 'к_Артикул ', 'к_Вид',
       'к_Заводской номер', 'к_Код', 'к_вес', 'к_розничная стоимость',
       'к_Родитель2корр', 'к_проба', 'к_Родитель3корр', 'к_Родитель4корр',
       'к_Родитель5корр', 'к_Вид номенклатуры', 'к_закупочная стоимость',
       'к_проблема', 'к_вставки', 'к_Склад_копия', 'дата запрса',
       'причина запроса на сверку - результат', 'Код', 'Штрихкод',
       'Вставки 1С', 'Вставки из ЮТД', 'gem_Вставки КОРРЕКТИРОВКИ',
       'Вставки 01-01-2023', 'Вставки корректировка',
       'дата корректировки вставки', 'корректировка вставки была',
       'Источник корректировки', 'Родитель1корр', 'Родитель2корр',
       'Родитель3корр', 'Родитель4корр', 'Родитель5корр', 'ut_Код',
       'ut_Старый код (Справочник "Номенклатура" (Общие))', 'ut_Склад',
       'ut_Склад.Склад / Магазин',
       'ut_Вставки (Справочник "Номенклатура" (Общие))', 'ut_Вид номенклатуры',
       'ut_Тип номенклатуры', 'ut_Номенклатура', 'ut_Группа

In [96]:
len(df_vst['Код'].unique()), len(df_vst)

(34269, 34269)

In [97]:
len(df_1c_res['Код'].unique()), len(df_1c_res)

(34269, 34269)

In [98]:
#Присоединяем к df_1c
df_1c_res = pd.merge(left=df_1c_res,
                     right=df_vst[['Код', 'Штрихкод', 'Вставки корректировка',
                 'дата корректировки вставки', 'корректировка вставки была','Источник корректировки',
                 'Родитель1корр', 'Родитель2корр', 'Родитель3корр', 'Родитель4корр', 'Родитель5корр']],
         how='left',
         right_on=['Код'],
         left_on=['Код']
         )

In [99]:
#223 пустых строк это по 3 договорам новые Валив и ДЯ
df_1c_res['Штрихкод'].isna().sum()

0

In [100]:
#df_1c_res.rename(columns={'Вставки':'Вставки010123'}, inplace=True)

# Анализ 0

In [101]:
df_1c_res.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 34269 entries, 0 to 34268
Data columns (total 43 columns):
 #   Column                                                                   Non-Null Count  Dtype         
---  ------                                                                   --------------  -----         
 0   Код                                                                      34269 non-null  object        
 1   Старый код (Справочник "Номенклатура" (Общие))                           28239 non-null  object        
 2   Склад                                                                    34269 non-null  object        
 3   Склад.Склад / Магазин                                                    34269 non-null  object        
 4   Вставки (Справочник "Номенклатура" (Общие))                              24298 non-null  object        
 5   Вид номенклатуры                                                         34269 non-null  object        
 6   Тип номенклату

In [102]:
df_1c_res.columns

Index(['Код', 'Старый код (Справочник "Номенклатура" (Общие))', 'Склад',
       'Склад.Склад / Магазин', 'Вставки (Справочник "Номенклатура" (Общие))',
       'Вид номенклатуры', 'Тип номенклатуры', 'Номенклатура', 'Группа',
       'Заводской номер (Справочник "Номенклатура" (Общие))',
       'Проба (Справочник "Номенклатура" (Общие))',
       'Цвет металла (Справочник "Номенклатура" (Общие))',
       'Дизайн (Справочник "Номенклатура" (Общие))', 'Артикул ',
       'Размерная группа Центрального камня (Справочник "Номенклатура" (Общие))',
       'Цвет Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Чистота Центрального Камня (Справочник "Номенклатура" (Общие))',
       'Размер (Справочник "Номенклатура" (Общие))', 'Наименование для печати',
       'Единица измерения веса',
       'Вес Вставок ПДК (Карат) (Справочник "Номенклатура" (Общие))',
       'Кол-во ДК (Справочник "Номенклатура" (Общие))',
       'Вес Вставок ДК (Карат) (Справочник "Номенклатура" (Общие))', 'Гру

In [103]:
for c in ['Родитель1корр', 'Родитель2корр', 'Родитель3корр', 'Родитель4корр', 'Родитель5корр']:
    print(c, '-->', df_1c_res.sample(10)[c])

Родитель1корр --> 10478    Ювелирные изделия
9475     Ювелирные изделия
10060    Ювелирные изделия
12547    Ювелирные изделия
20846    Ювелирные изделия
6163     Ювелирные изделия
19578    Ювелирные изделия
1455     Ювелирные изделия
25086    Ювелирные изделия
4780     Ювелирные изделия
Name: Родитель1корр, dtype: object
Родитель2корр --> 19309           Драгоценности Якутии ООО
21740    Возврат ООО "Невский бриллиант"
32755           Драгоценности Якутии ООО
24952           Драгоценности Якутии ООО
11956           Драгоценности Якутии ООО
3507            Драгоценности Якутии ООО
18784           Драгоценности Якутии ООО
6828                                МИЗО
32343           Драгоценности Якутии ООО
12050             Золото Якутии ОАО АПЮК
Name: Родитель2корр, dtype: object
Родитель3корр --> 17632    Sakha Minerals Limited SM21-01 от 28.09.2021
32992                    №34дм-10-01 от 01.10.2001 г.
752                            № 06-329/2 от 12.12.16
3621                     №10-121 о

In [104]:
df_1c_res.value_counts('Склад')

Склад
Ювелирные изделия               16506
Центральный                      1770
Иркутск                          1760
Ленина23                         1725
Пассаж                           1630
Кружало                          1583
Ленина, 28                       1551
Новосибирск                      1402
Хабаровск                        1336
Ленск                            1246
Мирный                           1140
Благовещенск                     1034
Интернет-магазин                  861
Выставочный                       647
Московское представительство       50
Склад "В пути"                     28
dtype: int64

На 15-10-2023
Склад
Ювелирные изделия               16506
Центральный                      1770
Иркутск                          1760
Ленина23                         1725
Пассаж                           1630
Кружало                          1583
Ленина, 28                       1551
Новосибирск                      1402
Хабаровск                        1336
Ленск                            1246
Мирный                           1140
Благовещенск                     1034
Интернет-магазин                  861
Выставочный                       647
Московское представительство       50
Склад "В пути"                     28
dtype: int64

на 19-09-2023 ут?
Склад
Ювелирные изделия               16505
Центральный                      1793
Иркутск                          1772
Ленина23                         1750
Пассаж                           1619
Ленина, 28                       1584
Кружало                          1564
Новосибирск                      1413
Хабаровск                        1343
Ленск                            1253
Мирный                           1156
Благовещенск                     1046
Интернет-магазин                  861
Выставочный                       818
Московское представительство       51
dtype: int64

# Запись результата в файл 2

In [105]:
os.listdir('../')

['1 Вводные',
 'del',
 'for_script03-11-2023',
 'for_script17-10-2023',
 'for_script24-10-2023',
 'ipynb',
 'ReadMe2023-10-18.docx',
 '~$adMe2023-09-19.docx',
 '~$adMe2023-10-18.docx',
 'Вставки Родитель1-5 коррекция',
 'Приложение к readme обнаружены несоответствия по номенклатуре 26-09-2023.xlsx',
 'Расчет03-11-2023',
 'Расчет17-10-2023',
 'Расчет24-10-2023',
 'Результаты анализа',
 'шаблон',
 'штрихкод_24-10-2023_16-33.xlsx']

In [106]:
os.mkdir('../for_script07-11-2023/')

In [108]:
folder_res = '../for_script07-11-2023/'
folder_res

'../for_script07-11-2023/'

In [109]:
os.listdir(folder_res)


[]

In [110]:
#os.mkdir(folder_res)

In [111]:
datetime.datetime.now().strftime('%d-%m-%Y_%H-%M')

'08-11-2023_14-17'

In [112]:
#запись промрезультата для проверки на excel
df_1c_res.to_excel(folder_res+f"выгрузка15-10-2023_все_склады_ДК_{datetime.datetime.now().strftime('%d-%m-%Y_%H-%M')}.xlsx")

In [113]:
os.listdir(folder_res)

['выгрузка15-10-2023_все_склады_ДК_08-11-2023_14-17.xlsx']

In [85]:
#df_1c_res.loc[df_1c_res['Код'] =='ЦБ-00040270','Сводка Вставки' ].to_list()

# del